[[ WidgetTouch ]] Add support for touch events in widgets

runrevmark opened the pull request:
This patch implements OnTouch(Start|Finish|Move|Cancel) in widgets.

The widget which hit-tests for the first touch in a sequence grabs
all touches in that sequence. The event handlers take no arguments, but instead the
full touch state is available through syntax. If the widget
becomes invisible, or is closed whilst active touches are present
Cancel is sent for all of them.

The currently active touch (the one which caused the event to
bubble) can be found by using ```the touch index```, and its
position using ```the touch position```.

The number of touches currently in progress can be found by using
```the number of touches``` and the position of a given touch using
```the position of touch tIndex```.

Note: Due to the way touch messages are currently processed above
the widget event system, widgets will receive mouse messages and
*then* touch messages. To make a widget which works on both Desktop
and Mobile, and wants touches on Mobile, the widget must have both
mouse and touch handlers, and check for platform in the mouse handlers.

runrevmark committed:
[[ WidgetTouch ]] Add support for touch events in widgets

This patch implements OnTouch(Start|Finish|Move|Cancel) in widgets.

The widget which hit-tests for the first touch in a sequence grabs
all touches in that sequence. The event handlers take no arguments, but instead the
full touch state is available through syntax. If the widget
becomes invisible, or is closed whilst active touches are present
Cancel is sent for all of them.

The currently active touch (the one which caused the event to
bubble) can be found by using ```the touch index```, and its
position using ```the touch position```.

The number of touches currently in progress can be found by using
```the number of touches``` and the position of a given touch using
```the position of touch tIndex```.

Note: Due to the way touch messages are currently processed above
the widget event system, widgets will receive mouse messages and
*then* touch messages. To make a widget which works on both Desktop
and Mobile, and wants touches on Mobile, the widget must have both
mouse and touch handlers, and check for platform in the mouse handlers.

montegoulding commented on engine/src/eventqueue.cpp:
```diff
@@ -1226,68 +1227,92 @@ static void handle_touch(MCStack *p_stack, MCEventTouchPhase p_phase, uint32_t p
-        if (t_target->gettype() == CT_WIDGET)
+    if (p_phase != kMCEventTouchPhaseBegan)
+    {
+        for(t_touch = s_touches; t_touch != nil; t_previous_touch = t_touch, t_touch = t_touch -> next)
```
@runrevmark there seems to be a rather large whitespace diff here. Do you want to drop that?

montegoulding commented on engine/src/widget-events.cpp:
```diff
@@ -496,26 +508,121 @@ void MCWidgetEventManager::event_paint(MCWidget *p_widget, MCGContextRef p_gcont
+    {
+        if (t_touched_widget == nullptr)
+        {
+            t_touched_widget = hitTest(p_widget->getwidget(), p_x, p_y);
```
@runrevmark hmm... so this means that multi-touch if the user is touching other objects only the first touch target will get events? I guess it makes sense seeing as the alternative is multiple objects thinking they are getting single touched

runrevmark replied:
Yes - that's correct. LCS isn't quite right at the moment - the control accepting the first touch is actually what should receive subsequent ones otherwise per-control multi-touch gestures won't work unless all touches hit-test within the widget (which I don't think is what is expected by a user).

montegoulding commented on engine/src/widget-events.cpp:
```diff
@@ -496,26 +508,121 @@ void MCWidgetEventManager::event_paint(MCWidget *p_widget, MCGContextRef p_gcont
+    }
+    else if (p_phase == kMCEventTouchPhaseBegan)
+    {
+        /* This shouldn't happen - it would mean we are getting two began
```
@runrevmark shouldn't we `MCUnreachable` this?

runrevmark replied:
I put that in because it was in the original code - we could put an MCAssert or such although this is a fault-tolerant resolution so I'm not sure it matters too much.

montegoulding commented on engine/src/widget-events.cpp:
```diff
@@ -496,26 +508,121 @@ void MCWidgetEventManager::event_paint(MCWidget *p_widget, MCGContextRef p_gcont
+    /* If we don't have a touched widget, then ignore the touch. */
+    if (t_touched_widget == nullptr)
+    {
+        return False;
```
@runrevmark so in this case I think the widget will just receive the messages in LCS. Should the target be changed to the underlying layer?

runrevmark replied:
The caller of event_touch has special code in it to handle widgets. If a widget is the target of the first touch, that touch-sequence is then bound to that widget (at the event dispatch step). This branch will only happen if it is not a begin phase, and 'cancel_touches' has been called in the event manager - in which case you don't want to take any further action.

montegoulding replied:
What I mean is because the `return` is only if `event_touch` is `true` [here](https://github.com/livecode/livecode/pull/6313/files#diff-7d875e990369273324bbf33f95262ce7R1297) then it will continue and send the LCS messages. Perhaps it should return there regardless of the return value of `event_touch`?

montegoulding commented:
@runrevmark do you think it would be worthwhile to add `the environment` as a helper or should we just write a function based on `the operating system`?

runrevmark commented:
@montegoulding I think that needs a little more thought - it might be better to add a utility function for detecting features of the platform and switch on whether the host platform 'supports touches' - that is what is the actual decidable thing here, not the platform or type of platform (mobile vs desktop) specifically.

montegoulding commented on engine/src/eventqueue.cpp:
```diff
@@ -1226,68 +1227,92 @@ static void handle_touch(MCStack *p_stack, MCEventTouchPhase p_phase, uint32_t p
-                case kMCEventTouchPhaseCancelled:
-                    t_target -> message_with_args(MCM_touch_release, p_id);
-                    break;
+                if (s_touches != nullptr)
```
@runrevmark Re my comment below perhaps we want:

```
if (t_target->gettype() == CT_WIDGET)
{
              if (MCwidgeteventmanager->event_touch(MCObjectCast<MCWidget>(t_target),
                                                  p_id, p_phase, t_touch_loc.x, t_touch_loc.y))
              {
                      if (s_touches != nullptr)
                      {
                            s_touches->is_widget = true;
                      }
              }
              return;
}
```

runrevmark replied:
Hmmm - yes - you are probably right. At the moment we don't do any 'fallthrough' of events from widgets to the host script object - which this would (as you point out) cause.

So, I think what we want to ensure here is that:
- if a widget hit-tests for the first touch in a sequence, then that widget gets *all* touches in that sequence (i.e. only the first touch is hit-tested)
- if a widget does not hit-test for the first touch in a sequence, then LCS gets *all* touches in that sequence

This is related to the issue of mouse messages not percolating for widgets if the widget doesn't handle them. I'll give that some thought - it might be we can use the same model for mouse as we come up with for touch messages. After all, ideally, if a widget is not interested in touch messages and is not going to do anything with them then LCS should probably get them - just as would be the ideal situation with mouse messages.

montegoulding replied:
Hmm... I have always thought that it would be good if just like we have a minimal set of properties that the engine handles for a widget (colors, font, rect etc) that a minimal set of messages would be a reasonable thing too. Then if a widget wants to notify the object of a specific event it uses a different message name. Not sure if that ship has sailed though.

montegoulding replied:
I think my concerns here may have been partly unfounded because after looking at `MCWidget::hittest` the widget won't be the target unless at least one of it or it's children is hit, however, we do have an issue that `m_touched_widget` can be reset each time there's a `kMCEventTouchPhaseBegan`. We should probably only try and set it if it's `nullptr` (first touch start) I think.

montegoulding replied:
One thing is if widgets grab the touch messages like this then you can't do something like:
- create 2 svg icon widgets
- create a behavior to handle touchMove to relocate the target

In the current implementation the second widget will not get any events and the first will only get them if the widget is scripted to send them. Perhaps what we really need is widgets that handle touch get all the touches once they grab them but LCS continues to send messages to the target which is determined for each event. I'm pretty sure that's what I'd like for mouse and key events too...

montegoulding commented:
@runrevmark I started doing some docs for this but I think there's a general issue with the use of touch index to identify touches. I believe the following will result two active touches with the same index:

1. touch with one finger
2. touch with a second finger
3. lift the first finger
4. touch again with the first finger

The only alternative would be to assign `m_index` to max active touch index but is there a reason we can't just drop `index` and use `id`?

montegoulding commented on engine/src/widget-syntax.cpp:
```diff
@@ -380,6 +380,63 @@ extern "C" MC_DLLEXPORT_DEF void MCWidgetGetClickCount(bool p_current, unsigned
+    r_count = MCwidgeteventmanager->GetTouchCount();
+}
+
+extern "C" MC_DLLEXPORT_DEF void MCWidgetGetPositionOfTouch(integer_t p_id, MCValueRef& r_point)
```
@runrevmark to reduce confusion (presuming we keep the `index` concept) we should rename this `p_index`

montegoulding commented on engine/src/widget.lcb:
```diff
@@ -642,6 +642,11 @@ public foreign handler MCWidgetGetClickPosition(in pCurrent as CBool, out rLocat
+public foreign handler MCWidgetGetTouchIndex(out rId as optional Integer) returns nothing binds to "<builtin>"
+public foreign handler MCWidgetGetTouchPosition(out rLocation as optional Point) returns nothing binds to "<builtin>"
+public foreign handler MCWidgetGetNumberOfTouches(out rCount as CUInt) returns nothing binds to "<builtin>"
+public foreign handler MCWidgetGetPositionOfTouch(in pId as Integer, out rLocation as optional Point) returns nothing binds to "<builtin>"
```
@runrevmark again rename `pId` here to `pIndex`

montegoulding commented on engine/src/widget.lcb:
```diff
@@ -750,6 +755,30 @@ begin
+end syntax
+
+syntax TheLocationOfTouch is prefix operator with property precedence
+    "the" "position" "of" "touch" <Id: Expression>
```
@runrevmark again here

montegoulding committed:
[[ WidgetTouch ]] Document touch API

montegoulding commented:
Ok my last commit here ensures that the new touch index is the current max active index + 1. This still means that within a multi-touch there may be more than one touch that is referred to by an index but I don't think this is much of an issue as long as we can't get two touches with the same index:

1. touch with one finer (index 1)
2. touch with second finger (index 2)
3. raise second finger (index 2 inactive)
4. touch with second finger (index 2 refers to new touch)

montegoulding commented:
Another alternative would be to increment the index for each touch in a multi-touch session and leave inactive touches around until the the last touch is complete. Then add:
```
the number of [active] touches
touch <index> is [not] active
```
This might be handy. For example on chrome on my phone lifting the first touch seems to disable all touches so to get that behavior it would be easy to guard touch handlers with:
```
if touch 1 is not active then
   return
end if
```

runrevmark commented:
@montegoulding : You are right - in a touch sequence, each separate touch should have a different index, each getting higher up. In your example sequence, I'd class the second touching of finger 2 as a separate touch so it should really have a second index.

Using the indexed array at least does mean that the indicies passed to the touch messages are ordered (i.e. higher indicies mean later touches), although I think any touch processing code has to keep around the index of the touches anyway in its own state so index is probably not that much different from id so...

I wonder if my attempt to use an indexed array here is flawed - at any moment in time the active touches are a set of triples (id, x, y) so we could consider 'the touches' to be an array mapping id to touch-info, the touch ids would be list of ids (i.e. the keys of that array).

The touch ids can still be reset to 0 for a new touch sequence, and we can keep the guarantee that a higher touch id means the touch came later than one with a lower touch id.

We could go with the top-level syntax of:
```
  the active touch - returns an array with id and position fields for the touch relevant to the current OnTouch* messages
  the touches - returns an array mapping id to touch array
  the number of touches - the number of elements in 'the touches'
```

Then any accessors we want to make things easier to code:
```
  the id of the active touch
  the id of touch <id>
  the position of the active touch
  the position of touch <id>
```
Essentially 'touch' becomes a chunk, using id as its index; and we have a related chunk 'the active touch'.

What do you think? This avoids having to have extra syntax or code to iterate through the number of touches and then check for activeness.

The main thing I was trying to achieve by using an indexed array was to make it slightly easier to detect the first touch and such - but after your comments I think that was probably a 'bad' way to go beyond keeping the following slightly nicer properties that such an idea had:

- when a touch sequence ends, the id is reset to 0 (i.e. the first touch in a sequence always has id 1)
- touches don't reuse previous ids in a sequence
- a touch which occurred later will have a higher id

What do you think?

montegoulding commented:
I think I like:
```
the touch id -> id of the current touch
the touch position -> position of the current touch
the touches -> map of id -> position
the number of touches -> count of active touches
the position of touch <id> -> position of specific touch
```
where `id` is a sequence incremented per touch and reset to 0 whenever the number of touches returns to 0

montegoulding commented:
@runrevmark actually I'm going to do `the touch ids` rather than `the touches` because we don't have an integer keyed map in LCB it will actually be really annoying to return an array... really wish we had a `Map<key type,value type>` thing we could use.

montegoulding committed:
[[ WidgetTouch ]] Change touch ID to sequence

This patch changes touch ID to be a sequence incremented
for each touch in a set of multi-touch events. The sequence
is reset when there are no longer any active touches so that
the first touch is always id 1 and each touch increments.

This patch also implements `the touch ids` which returns a list
of currently active touch IDs.

runrevmark committed:
[[ WidgetTouch ]] Implement default (script) behavior for touch

This patch changes the way touch messages are handled when dealing
with widgets.

Touch messages are now only sent to a widget if it, or any of it
widget ancestors have at least one OnTouch* event handler.

If the first touch in a sequence is handled by a widget, then all
subsequent touches in that sequence will be grabbed by the widget
event manager.

If the first touch in a sequence is not handled by a widget, then
all subsequent touches in that sequence will be grabbed by the
engine.

montegoulding committed:
[[ WidgetTouch ]] Add release note

montegoulding committed:
[[ WidgetTouch ]] Remove errant whitespace

montegoulding commented on engine/src/eventqueue.cpp:
```diff
@@ -1273,7 +1273,7 @@ static void handle_touch(MCStack *p_stack, MCEventTouchPhase p_phase, uint32_t p
         t_touch -> next = s_touches;
         t_touch -> id = p_id;
-        
+
```
@runrevmark Somehow we got a backspace in here `U+0008` 🙀

montegoulding committed:
[[ WidgetTouch ]] Return true on success

livecodeali commented on engine/src/widget-events.cpp:
```diff
@@ -496,26 +508,121 @@ void MCWidgetEventManager::event_paint(MCWidget *p_widget, MCGContextRef p_gcont
 }
 
-void MCWidgetEventManager::event_touch(MCWidget* p_widget, uint32_t p_id, MCEventTouchPhase p_phase, int2 p_x, int2 p_y)
+Bool MCWidgetEventManager::event_touch(MCWidget* p_widget, uint32_t p_id, MCEventTouchPhase p_phase, int2 p_x, int2 p_y)
```
@runrevmark Any particular reason why this is a `Bool`?

montegoulding replied:
Our style guide doesn't mention a preference but previously implemented methods of this class use `Boolean` and of the new ones this is the only one that uses `Bool`. I will change to `bool` so at least this patch is consistent.

livecodeali commented on engine/src/widget-events.cpp:
```diff
@@ -919,199 +1067,59 @@ bool MCWidgetEventManager::keyUp(MCWidgetRef p_widget, MCStringRef p_string, Key
-        mouseMove(p_widget, p_x, p_y);
-        mouseRelease(p_widget, 1);
+        // No empty slots found. Extend the array.
+        i = m_touches.Size();
```
This line is redundant

livecodeali commented:
This looks good to me - can we change where applicable all the occurrences of Index to ID (particularly in the syntax name, as that will appear in the docs)

montegoulding committed:
[[ WidgetTouch ]] Remove redundant line

montegoulding committed:
[[ WidgetTouch ]] Use bool instead of Bool

montegoulding committed:
[[ WidgetTouch ]] Replace index with id

montegoulding committed:
[[ WidgetTouch ]] Set touch target to widget

montegoulding committed:
[[ WidgetTouch ]] Increment touch count

montegoulding committed:
[[ WidgetTouch ]] Ensure touch target is valid

montegoulding commented:
@livecodeali I have now resolved the crashes I was seeing on this so it's ready for review

livecodeali commented on engine/src/eventqueue.cpp:
```diff
@@ -1281,16 +1281,21 @@ static void handle_touch(MCStack *p_stack, MCEventTouchPhase p_phase, uint32_t p
         else
         {
             t_target = s_touches -> target;
+            t_touch -> target = t_target -> GetHandle();
```
Do we know that `s_touches->target` hasn't been deleted at any point?

montegoulding replied:
Yes it shouldn't be possible for `s_touches_for_widget` to be true and `s_touches` to be `nullptr`. I had considered adding `MCAssert` here and I guess that will clarify the point. What do you think?

livecodeali replied:
I think `GetHandle` will assert anyway in this case won't it? So probably fine!

montegoulding replied:
hehe... well it will definitely crash with hopefully a reasonable message ;-)

livecodeali commented:
@livecode-vulcan review ok fc310fa

livecode-vulcan commented:
💙 review by @livecodeali ok fc310fac292f8b9568c80a7f0af37e6c730d0252

livecode-vulcan commented:
😎 test success fc310fac292f8b9568c80a7f0af37e6c730d0252

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1214)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1230)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1204)
* try-community-universal-ios-iphoneos11.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.2/builds/157)
* try-community-universal-ios-iphonesimulator11.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.2/builds/157)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1209)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1235)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1214)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1230)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1213)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1223)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1263)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1208)
